apps/sms/services/js/worker_client.js

```diff
@@ -0,0 +1,215 @@
+ streamClient
+ */
+
+(function(exports) {
+  'use strict';
+
+  const debug = 1 ?
+    (arg1, ...args) => console.log(`[WorkerClient] ${arg1}`, ...args):
+    () => {};
+
+  /**
+   * Supported worker types.
+   * @enum {string}
+   */
+  const WORKER_TYPE = Object.freeze({
+    WORKER: 'worker',
+    SHARED_WORKER: 'shared-worker'
+  });
+
+  /**
+   * Supported lifetime strategies or the way we manage workers.
+   * @enum {string}
+   */
+  const LIFETIME_STRATEGY = Object.freeze({
+    ALWAYS_ALIVE: 'always-alive',
+    ALIVE_WHEN_VISIBLE: 'alive-when-visible'
+  });
+
+  /**
+   * Default values for the options supported by WorkerClient.
+   * @type {Object}
+   */
+  const DEFAULT_OPTIONS = Object.freeze({
+    /**
+     * Type of the worker to use to run service.
+     * @type {WORKER_TYPE}
+     */
+    type: WORKER_TYPE.SHARED_WORKER,
+
+    /**
+     * Lifetime strategy to use.
+     * @type {LIFETIME_STRATEGY}
+     */
+    lifetime: LIFETIME_STRATEGY.ALWAYS_ALIVE,
+
+    /**
+     * Indicates that both client and service support "streams", so that we
+     * should load appropriate bridge plugin.
+     * @type {boolean}
+     */
+    supportStreams: false,
+
+    /**
+     * Indicates if we want to run worker with the service immediately,
+     * alternatively service will be run only on the first demand.
+     * @type {boolean}
+     */
+    preRun: false,
+
+    /**
+     * Sets the number of milliseconds we should wait before actually shutting
+     * down client once shut down has been requested.
+     */
+    disposeTimeout: 3000,
+
+    /**
+     * Default client-service response timeout. If "false" - then client request
+     * will never be time outed.
+     * @type {Boolean|Number}
+     */
+    timeout: false
+  });
+
+  const priv = Object.freeze({
+    options: Symbol('options'),
+
+    client: Symbol('client'),
+    endpoint: Symbol('endpoint'),
+
+    create: Symbol('create')
+  });
+
+  const WorkerClient = function(serviceName, servicePath, options = {}) {
+    debug('Initialized');
+
+    if (!serviceName || !servicePath) {
+      throw new Error('Both service name and path are required!');
+    }
+
+    this[priv.options] = Object.assign(
+      { serviceName, servicePath }, DEFAULT_OPTIONS, options
+    );
+
+    if (options.lifetime === LIFETIME_STRATEGY.ALIVE_WHEN_VISIBLE) {
+      document.addEventListener('visibilitychange', () => {
+        if (document.hidden) {
+          debug('Document is hidden');
+
+          //this.dispose();
+        }
+      });
+    }
+  };
+
+  WorkerClient.WORKER_TYPE = WORKER_TYPE;
+  WorkerClient.LIFETIME_STRATEGY = LIFETIME_STRATEGY;
+
+  WorkerClient.prototype = {
+    /**
+     * Bridge client instance.
+     * @type {Client}
+     * @private
+     */
+    [priv.client]: null,
+
+    /**
+     * Service endpoint instance (either Worker or SharedWorker).
+     * @type {Worker|SharedWorker}
+     * @private
+     */
+    [priv.endpoint]: null,
+
+    method(...args) {
```

Comment:
Here I have several options as well:
* Current: consumer of ```WorkerClient``` will have Promise-based ```method``` and ```stream``` (I'd likely use Proxy to expose all bridge methods in single way if we decide use this approach) , but ```stream``` on bridge client isn't Promise based. So I don't really like that BridgeClient and WorkerClient will have different API.
* We don't expose these method and consumer does smth like ```WorkerClient.getClient().then((bridgeClient) => bridge.client.method(.....))``` But here I don't like that bridge client is exposed to consumer...

What do you think?